### PR TITLE
Handle messages and notifications

### PR DESCRIPTION
**Pull request by cpacia**

### CODE REVIEW

File: db/datastore.py [hunk 1]
```diff
@@ -275,3 +275,39 @@ def get_followers(self):
...
+
+    def get_messages(self, guid, message_type):
+        cursor = self.db.cursor()
+        cursor.execute('''SELECT guid, signed_pubkey, encryption_pubkey, subject, message_type, message,
+timestamp, signature FROM messages WHERE guid=? AND message_type=?''', (guid, message_type))
+        ret = cursor.fetchall()
+        if not ret:
+            return None
+        else:
+            return ret
```
Comment (hoffmabc) [inline db/datastore.py]: return not ret ? None : ret

This looks so much crazier. I just wanted to put this here.

File: db/datastore.py [hunk 2]
```diff
@@ -275,3 +275,39 @@ def get_followers(self):
...
+
+    def delete_message(self, guid):
+        cursor = self.db.cursor()
+        cursor.execute('''DELETE FROM messages WHERE guid=? AND message_type="CHAT"''', (guid, ))
```
Comment (hoffmabc) [inline db/datastore.py]: Why not have a flag column that's called 'archived' and just flag the rows you "deleted". That way we can support Undo.

File: market/listeners.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
+__author__ = 'chris'
```
Comment (hoffmabc) [inline market/listeners.py]: Why don't you ever put your whole name or GitHub handle?

File: market/listeners.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    implements(MessageListener)
+
+    def __init__(self, web_socket_factory):
+        self.ws = web_socket_factory
+        self.db = MessageStore()
+
+    def notify(self, sender_guid, signed_pubkey, encryption_pubkey,
+               subject, message_type, message, timestamp, signature):
+
+        self.db.save_message(sender_guid, signed_pubkey, encryption_pubkey, subject,
```
Comment (hoffmabc) [inline market/listeners.py]: Maybe we should make a Message object rather than pass in an insane amount of parameters?

File: openbazaard.py [hunk 5]
```diff
@@ -74,4 +71,13 @@ def notify(sender_guid, encryption_pubkey, subject, message_type, message):
 factory = jsonrpc.RPCFactory(RPCCalls(kserver, mserver, keys))
 reactor.listenTCP(18465, factory, interface="127.0.0.1")
 
+# web sockets
+ws_factory = WSFactory("ws://127.0.0.1:18466", mserver)
```
Comment (hoffmabc) [inline openbazaard.py]: Let's make sure to move this to a constants file.

**Comment by hoffmabc** [comment]
Obviously some pylint issues to clean up, but won't dwell since you marked [WIP].

File: db/datastore.py (hunk 2 above)
Comment (gubatron) [inline db/datastore.py]: as a devil's advocate, here's an alternative.

1. Re: Why? Privacy and being able to certainly delete something you don't want on records is smart.
2. You can have the actual deletion done unless Undo has been clicked, say 15 seconds later, and have the option to back out anytime before that happens.

File: db/datastore.py [hunk 6]
```diff
@@ -275,3 +275,39 @@ def get_followers(self):
...
+        self.db = lite.connect(DATABASE)
+        self.db.text_factory = str
+        try:
+            cursor = self.db.cursor()
+            cursor.execute('''CREATE TABLE messages(guid BLOB , signed_pubkey BLOB,
+encryption_pubkey BLOB, subject TEXT, message_type TEXT, message TEXT, timestamp, INTEGER, signature BLOB)''')
+            cursor.execute('''CREATE INDEX idx1 ON messages(guid);''')
+            self.db.commit()
+        except Exception:
+            pass
```
Comment (gubatron) [inline db/datastore.py]: perhaps keeping a state here on which you know that the table couldn't be created might help.

Comment (gubatron) [inline db/datastore.py]: not sure where this constructor gets invoked, but does this table need to be created everytime you instatiate `MessageStore`?

File: market/listeners.py [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+from zope.interface import implements
+from db.datastore import MessageStore
+from twisted.internet import task, reactor
+
+class MessageListenerImpl(object):
+    implements(MessageListener)
+
+    def __init__(self, web_socket_factory):
+        self.ws = web_socket_factory
+        self.db = MessageStore()
```
Comment (gubatron) [inline market/listeners.py]: hmm, so here I suppose. I guess you silently want it to fail on every case after the time it's first created.

File: market/listeners.py (hunk 3 above)
Comment (cpacia) [inline market/listeners.py]: That's just autogenerated and I forget to change it. I can go back and change them later.

File: market/listeners.py (hunk 4 above)
Comment (cpacia) [inline market/listeners.py]: I could just pass in the protobuf object. I'd still need to parse it in the `notify` function though.

File: db/datastore.py (hunk 6 above)
Comment (cpacia) [inline db/datastore.py]: It should just pass over it if the table already exists.

Comment (gubatron) [inline db/datastore.py]: it's just a bit weird seing such a defining db operation lightly used on a constructor.
The api user might not be aware, might try to invoke MessageStore on different threads, etc.
I'd def. move that logic to a method that describes that action.

Is there another place where all tables are created? I thought there was some class that happens on the first OB run where all tables are created. Perhaps that's a better place for this, also easier to maintain "where the hell is that table created?" kind of thing going through a maintainer's mind in the future.

Comment (hoffmabc) [inline db/datastore.py]: I think this goes back to the talk about db schema creation and also migrations going forward. Maybe we should extract this out to some db specific stuff.

Comment (cpacia) [inline db/datastore.py]: Yeah that sounds like a good idea

**Pull request merged**